src/zarr/codecs/arrow.py

```diff
@@ -0,0 +1,81 @@
+from __future__ import annotations
+
+from dataclasses import dataclass
+from typing import TYPE_CHECKING
+
+import pyarrow as pa
+
+from zarr.abc.codec import ArrayBytesCodec
+from zarr.array_spec import ArraySpec
+from zarr.buffer import Buffer, NDBuffer
+from zarr.codecs.registry import register_codec
+from zarr.common import JSON, parse_named_configuration
+
+if TYPE_CHECKING:
+    from typing_extensions import Self
+
+CHUNK_FIELD_NAME = "zarr_chunk"
+
+
+@dataclass(frozen=True)
+class ArrowRecordBatchCodec(ArrayBytesCodec):
+    def __init__(self) -> None:
+        pass
+
+    @classmethod
+    def from_dict(cls, data: dict[str, JSON]) -> Self:
+        _, configuration_parsed = parse_named_configuration(
+            data, "arrow", require_configuration=False
+        )
+        configuration_parsed = configuration_parsed or {}
+        return cls(**configuration_parsed)
+
+    def to_dict(self) -> dict[str, JSON]:
+        return {"name": "arrow"}
+
+    def evolve_from_array_spec(self, array_spec: ArraySpec) -> Self:
+        return self
+
+    async def _decode_single(
+        self,
+        chunk_bytes: Buffer,
+        chunk_spec: ArraySpec,
+    ) -> NDBuffer:
+        assert isinstance(chunk_bytes, Buffer)
+
+        # TODO: make this compatible with buffer prototype
+        arrow_buffer = memoryview(chunk_bytes.to_bytes())
+        with pa.ipc.open_stream(arrow_buffer) as reader:
+            batches = [b for b in reader]
+        assert len(batches) == 1
+        arrow_array = batches[0][CHUNK_FIELD_NAME]
+        chunk_array = chunk_spec.prototype.nd_buffer.from_ndarray_like(
+            arrow_array.to_numpy(zero_copy_only=False)
+        )
+
+        # ensure correct chunk shape
+        if chunk_array.shape != chunk_spec.shape:
+            chunk_array = chunk_array.reshape(
+                chunk_spec.shape,
+            )
+        return chunk_array
+
+    async def _encode_single(
+        self,
+        chunk_array: NDBuffer,
+        chunk_spec: ArraySpec,
+    ) -> Buffer | None:
+        assert isinstance(chunk_array, NDBuffer)
+        arrow_array = pa.array(chunk_array.as_ndarray_like().ravel())
```

Comment:
Should probably be `chunk_array.as_numpy_array()` here since `pa.array()` doesn't recognize CuPy arrays? Would be good to add a GPU test here for safety.

In theory, it would be possible to do zero-copy transfers for CuPy arrays too but would need to go from [CuPy -> Numba](https://numba.readthedocs.io/en/stable/cuda/cuda_array_interface.html#numba.cuda.as_cuda_array) first and then [Numba -> Arrow](https://arrow.apache.org/docs/python/integration/cuda.html#numba-to-arrow).